### PR TITLE
DRILL-7356: Introduce session options for the Drill Metastore

### PR DESCRIPTION
**Pull request by vvysotskyi**
For details please see [DRILL-7356](https://issues.apache.org/jira/browse/DRILL-7356).

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 1]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
   public static final RangeLongValidator QUERY_MAX_ROWS_VALIDATOR = new RangeLongValidator(QUERY_MAX_ROWS, 0, Integer.MAX_VALUE,
       new OptionDescription("The maximum number of rows that the query will return. This can be only set at a SYSTEM level by an admin. (Drill 1.16+)"));
 
+  /**
+   * Option that enables Drill metastore usage.
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * Option that enables Drill Metastore usage.
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 2]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+   */
+  public static final String METASTORE_ENABLED = "metastore.enabled";
+  public static final BooleanValidator METASTORE_ENABLED_VALIDATOR = new BooleanValidator(METASTORE_ENABLED,
+      new OptionDescription("Enables Drill metastore usage to be able to store table metadata " +
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
      new OptionDescription("Enables Drill Metastore usage to be able to store table metadata " +
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 3]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+          "during ANALYZE TABLE commands execution and to be able to read table metadata during regular " +
+          "queries execution or when querying some INFORMATION_SCHEMA tables. " +
+          "This option is not active for now. Default is false. (Drill 1.17+)"));
+
+  /**
+   * Option for specifying maximum level depth for collecting metadata
+   * which will be stored in the Drill metastore. For example, when {@code FILE} value
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * which will be stored in the Drill Metastore. For example, when {@code FILE} value
```

Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * which will be stored in the Drill metastore. For example, when {@code FILE} level value
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 4]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+   * is set, {@code ROW_GROUP} level metadata wouldn't be collected and stored into the metastore.
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * is set, {@code ROW_GROUP} level metadata eon't be collected and stored into the Metastore.
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 5]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+  /**
+   * Option for enabling usage of the schema stored to the metastore.
+   */
+  public static final String METASTORE_USE_SCHEMA_METADATA = "metastore.metadata.use_schema";
+  public static final BooleanValidator METASTORE_USE_SCHEMA_METADATA_VALIDATOR = new BooleanValidator(METASTORE_USE_SCHEMA_METADATA,
+      new OptionDescription("Enables usage of the schema stored to the metastore. " +
+          "This option is not active for now. Default is false. (Drill 1.17+)"));
+
+  /**
+   * Option for enabling usage of the statistics stored to the metastore at the planning stage.
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * Option for enabling statistics usage, stored in the Metastore, at the planning stage.
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 6]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+   */
+  public static final String METASTORE_USE_STATISTICS_METADATA = "metastore.metadata.use_statistics";
+  public static final BooleanValidator METASTORE_USE_STATISTICS_METADATA_VALIDATOR = new BooleanValidator(METASTORE_USE_STATISTICS_METADATA,
+      new OptionDescription("Enables usage of the statistics stored to the metastore at the planning stage. " +
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]: Same here.

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 7]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+   */
+  public static final String METASTORE_CTAS_AUTO_COLLECT_METADATA = "metastore.metadata.ctas.auto-collect";
+  public static final EnumeratedStringValidator METASTORE_CTAS_AUTO_COLLECT_METADATA_VALIDATOR = new EnumeratedStringValidator(METASTORE_CTAS_AUTO_COLLECT_METADATA,
+      new OptionDescription("Specifies whether schema and / or column statistics will be " +
+          "automatically collected for every table after CTAS and CTTAS. " +
+          "This option is not active for now. Default is 'NONE'. (Drill 1.17+)"),
+      "NONE", "ALL", "SCHEMA");
+
+  /**
+   * Option for allowing using file metadata cache if required metadata is absent in the metastore.
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
   * Option for allowing using file metadata cache if required metadata is absent in the Metastore.
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 8]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+   */
+  public static final String METASTORE_FALLBACK_TO_FILE_METADATA = "metastore.metadata.fallback_to_file_metadata";
+  public static final BooleanValidator METASTORE_FALLBACK_TO_FILE_METADATA_VALIDATOR = new BooleanValidator(METASTORE_FALLBACK_TO_FILE_METADATA,
+      new OptionDescription("Allows using file metadata cache for the case when required metadata is absent in the metastore. " +
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
      new OptionDescription("Allows using file metadata cache for the case when required metadata is absent in the Metastore. " +
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 9]
```diff
@@ -1044,4 +1044,67 @@ public static String bootDefaultFor(String name) {
...
+          "This option is not active for now. Default is true. (Drill 1.17+)"));
+
+  /**
+   * Option for specifying the number of attempts for retrying query planning after detecting that query metadata is changed.
+   */
+  public static final String METASTORE_RETRIVAL_RETRY_ATTEMPTS = "metastore.retrival.retry_attempts";
+  public static final IntegerValidator METASTORE_RETRIVAL_RETRY_ATTEMPTS_VALIDATOR = new IntegerValidator(METASTORE_RETRIVAL_RETRY_ATTEMPTS,
+      new OptionDescription("Specifies the number of attempts for retrying query planning after detecting that query metadata is changed. " +
+          "If the number of retries was exceeded, query will be planned without metadata information from Metastore. " +
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]:
```suggestion
          "If the number of retries was exceeded, query will be planned without metadata information from the Metastore. " +
```

File: exec/java-exec/src/main/resources/drill-module.conf [hunk 10]
```diff
@@ -691,5 +691,13 @@ drill.exec.options: {
     # ========= rm related options ===========
     exec.rm.queryTags: "",
     exec.rm.queues.wait_for_preferred_nodes: true,
-    exec.statistics.tdigest_compression: 100
+    exec.statistics.tdigest_compression: 100,
+    # ========= metastore related options ===========
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/resources/drill-module.conf]:
```suggestion
    # ========= Metastore related options ===========
```

**Comment by vvysotskyi** [comment]
@arina-ielchiieva, thanks for the review, I have addressed CR comments. Please take a look again.

**Comment by arina-ielchiieva** [comment]
+1

**Pull request closed without merging**